src/main/java/com/aws/greengrass/clientdevices/auth/util/WildcardTrie.java

```diff
@@ -0,0 +1,334 @@
+ */
+public class WildcardTrie {
+    protected static final String GLOB_WILDCARD = "*";
+    protected static final String MQTT_MULTILEVEL_WILDCARD = "#";
+    protected static final String MQTT_SINGLELEVEL_WILDCARD = "+";
+    protected static final String MQTT_LEVEL_SEPARATOR = "/";
+    protected static final char nullChar = '\0';
+    protected static final char escapeChar = '$';
+    protected static final char wildcardChar = GLOB_WILDCARD.charAt(0);
+    protected static final char multiLevelWildcardChar = MQTT_MULTILEVEL_WILDCARD.charAt(0);
+    protected static final char singleLevelWildcardChar = MQTT_SINGLELEVEL_WILDCARD.charAt(0);
+    protected static final char levelSeparatorChar = MQTT_LEVEL_SEPARATOR.charAt(0);
+
+
+    private boolean isTerminal;
+    private boolean isTerminalLevel;
+    private boolean isWildcard;
+    private boolean isMQTTWildcard;
+    private boolean matchAll;
+    private final Map<String, WildcardTrie> children =
+            new DefaultConcurrentHashMap<>(WildcardTrie::new);
+
+    public void set(String subject) {
+        children.clear();
+        add(subject);
+    }
+
+    /**
+     * Add allowed resources for a particular operation.
+     * - A new node is created for every occurrence of a wildcard (*, #, +).
+     * - Only nodes with valid usage of wildcards are marked with isWildcard or isMQTTWildcard.
+     * - Any other characters are grouped together to form a node.
+     * - Just a '*' or '#' creates a Node setting matchAll to true and would match all resources
+     *
+     * @param subject resource pattern
+     */
+    public void add(String subject) {
+        if (subject == null) {
+            return;
+        }
+        if (subject.equals(GLOB_WILDCARD)) {
+            WildcardTrie initial = this.children.get(GLOB_WILDCARD);
+            initial.matchAll = true;
+            initial.isTerminal = true;
+            initial.isWildcard = true;
+            return;
+        }
+        if (subject.equals(MQTT_MULTILEVEL_WILDCARD)) {
+            WildcardTrie initial = this.children.get(MQTT_MULTILEVEL_WILDCARD);
+            initial.matchAll = true;
+            initial.isTerminal = true;
+            initial.isMQTTWildcard = true;
+            return;
+        }
+        if (subject.equals(MQTT_SINGLELEVEL_WILDCARD)) {
+            WildcardTrie initial = this.children.get(MQTT_SINGLELEVEL_WILDCARD);
+            initial.isTerminal = true;
+            initial.isMQTTWildcard = true;
+            return;
+        }
+        if (subject.startsWith(MQTT_SINGLELEVEL_WILDCARD + MQTT_LEVEL_SEPARATOR)) {
+            WildcardTrie initial = this.children.get(MQTT_SINGLELEVEL_WILDCARD);
+            initial.isMQTTWildcard = true;
+            initial.add(subject.substring(1), true);
+        }
+
+        add(subject, true);
+    }
+
+    @SuppressWarnings("PMD.AvoidDeeplyNestedIfStmts")
+    private WildcardTrie add(String subject, boolean isTerminal) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.